File: 10-01-iterator-graph/Main.java
import java.util.Iterator;

import adts.LinkedList;
import adts.List;

public class Main {
  public static void main(String[] args) {
    List<String> l = new LinkedList<>();

    l.insert("a");
    l.insert("b");
    l.insert("c");
    l.insert("d");
    l.insert("e");
    l.insert("f");
    l.insert("g");
    l.insert("h");

    // O(N^2)
    for (int i = 0; i < l.size(); i++) {
      System.out.println(l.get(i));
    }

    // O(N)
    Iterator<String> it = l.iterator();
    while (it.hasNext()) {
      String s = it.next();
      System.out.println(s);
    }

    // O(N)
    for (String s : l) {
      System.out.println(s);
    }
  }
}

File: 10-01-iterator-graph/main.cpp
#include "adts/linked_list.cpp"
#include <iostream>

int main()
{
  list<int> *l = new linked_list<int>();

  for (int i = 0; i < 10; i++)
  {
    l->insert(i);
  }

  iterator<int> *it = l->iterator();

  while (it->hasNext())
  {
    int n = it->next();
    std::cout << n << std::endl;
  }
}
